collector/ethtool_linux.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !noethtool
+
+// The majority of this code is derived from https://github.com/safchain/ethtool/blob/master/ethtool.go
+// By Sylvain Afchain. Used under the Apache license.
+
+package collector
+
+import (
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs/sysfs"
+	"github.com/safchain/ethtool"
+)
+
+type ethtoolCollector struct {
+	fs      sysfs.FS
+	entries *prometheus.Desc
+	logger  log.Logger
+}
+
+func init() {
+	registerCollector("ethtool", defaultDisabled, NewEthtoolCollector)
+}
+
+// NewEthtoolCollector returns a new Collector exposing ethtool stats.
+func NewEthtoolCollector(logger log.Logger) (Collector, error) {
+	fs, err := sysfs.NewFS(*sysPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open sysfs: %w", err)
+	}
+
+	return &ethtoolCollector{
+		fs: fs,
+		entries: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "ethtool", "stats"),
+			"Network interface card statistics",
+			[]string{"device", "stat"}, nil,
+		),
+		logger: logger,
+	}, nil
+}
+
+func (c *ethtoolCollector) Update(ch chan<- prometheus.Metric) error {
+	netClass, _ := c.fs.NetClass()
+
+	for device := range netClass {
+		stats, err := ethtool.Stats(device)
+		if err == nil {
+			// Suppressing errors because it's hard to tell what interfaces support ethtool and which don't.
```

Comment:
The Go style here is to use a `continue`, rather than increase the indentation.

```suggestion
		if err == nil {
			// Suppressing errors because it's hard to tell what interfaces support ethtool and which don't.
```
```suggestion
		if err != nil {
			// Suppressing errors because it's hard to tell what interfaces support ethtool and which don't.
			continue
		}
```